### ISSUE TITLE
[BUG] VM got IO error after host restart

### ISSUE THREAD

**Issue by FrankYang0529** [harvester/harvester#4633]
**Describe the bug**
VM can't restart after a host accidentally restarts. VMI changes frequently between ready and unready and it has `VMI was paused, IO error` error message. At first, related Volume's robustness is fault and engine is stopped. After around 30 minutes, the volume is attached and engine is running, but VMI still can't get ready. We have to check why VM can't be ready after the LH engine is back.

```
> kubectl get vmi -w
NAME                          AGE   PHASE     IP             NODENAME           READY
rke2-3-pool1-47f7bd68-r99dx   44m   Running                  harvester-node-2   False
rke2-3-pool1-47f7bd68-r99dx   44m   Running   192.168.3.56   harvester-node-2   False
rke2-3-pool1-47f7bd68-r99dx   44m   Running                  harvester-node-2   False
rke2-3-pool1-47f7bd68-r99dx   44m   Running                  harvester-node-2   True
rke2-3-pool1-47f7bd68-r99dx   44m   Running                  harvester-node-2   False
rke2-3-pool1-47f7bd68-r99dx   44m   Running   192.168.3.56   harvester-node-2   False
rke2-3-pool1-47f7bd68-r99dx   44m   Running   192.168.3.56   harvester-node-2   True
rke2-3-pool1-47f7bd68-r99dx   44m   Running                  harvester-node-2   True
rke2-3-pool1-47f7bd68-r99dx   44m   Running                  harvester-node-2   False
```

**To Reproduce**
1. Create a 3-node harvester cluster.
2. Provision a 3-node RKE2 cluster on the harvester and wait for it's ready.
3. Choose one node to stop kubelet.
    a. Find the pid of kubelet process by `ps -aux | grep kubelet`.
    b. Freeze the kubelet process by `kill -STOP <kubelet-pid>`. 
4. After the node is unavailable. Run `shutdown now -r`.
5. After the node is running again, the VM can't be ready.

**Expected behavior**
The RKE2 VM can be migrated to other hosts and the VM can be ready after a while.

**Support bundle**
[supportbundle_f3b2af65-e81b-4c19-939a-14ff8688c5a1_2023-10-19T03-37-52Z.zip](https://github.com/harvester/harvester/files/13037503/supportbundle_f3b2af65-e81b-4c19-939a-14ff8688c5a1_2023-10-19T03-37-52Z.zip)

**Environment**
 - Harvester ISO version: v1.2.0
 - Underlying Infrastructure (e.g. Baremetal with Dell PowerEdge R630): KVM

**Additional context**
Related LH issues:
https://github.com/longhorn/longhorn/issues/6919
https://github.com/longhorn/longhorn/issues/6961

**Work around**
Restart the failed VM and it can get back and be ready after rescheduling.

**Comment by FrankYang0529** [comment]
Updated reproduce steps and work around. New steps are more reproducible. I tried six times and I can get a unready VM in four times.

From VM scheduling perspective, we can try to restart VM if it matches following conditions:
* PVC is bound and related Volume is healthy.
* VM pod is running but there is `ReadinessGatesNotReady` in the condition.
* VMI has `PausedIOError`.

**Comment by FrankYang0529** [comment]
New support bundle. In this bundle, the volume `pvc-2c649b39-95ef-474a-97fb-2b8e74bfa8d7` is healthy, but the VMI `rke2-2-pool1-926e1b37-fx2bl` can't be ready.
[supportbundle_aed7c5f9-39d1-4c1d-8fa8-b1a09baceea4_2023-10-20T03-08-34Z.zip](https://github.com/harvester/harvester/files/13049641/supportbundle_aed7c5f9-39d1-4c1d-8fa8-b1a09baceea4_2023-10-20T03-08-34Z.zip)

**Comment by FrankYang0529** [comment]
If we check log in VM pod, we can see it keeps running `Synced vmi` with IO Error.

```
{"component":"virt-launcher","kind":"","level":"info","msg":"Synced vmi","name":"rke2-2-pool1-926e1b37-fx2bl","namespace":"default","pos":"server.go:190","timestamp":"2023-1
0-20T06:42:25.787009Z","uid":"feeaed37-53a7-4c61-bedd-d56b2f57e61e"}
{"component":"virt-launcher","level":"info","msg":"Domain name event: default_rke2-2-pool1-926e1b37-fx2bl","pos":"client.go:413","timestamp":"2023-10-20T06:42:25.821255Z"}
{"component":"virt-launcher","level":"info","msg":"kubevirt domain status: Paused(3):IOError(5)","pos":"client.go:288","timestamp":"2023-10-20T06:42:25.823399Z"}
{"component":"virt-launcher","level":"info","msg":"DomainLifecycle event 4 with reason 0 received","pos":"client.go:435","timestamp":"2023-10-20T06:42:25.833676Z"}
{"component":"virt-launcher","level":"info","msg":"DomainLifecycle event 3 with reason 2 received","pos":"client.go:435","timestamp":"2023-10-20T06:42:25.852570Z"}
{"component":"virt-launcher","level":"info","msg":"Libvirt event channel is full, dropping event.","pos":"client.go:443","timestamp":"2023-10-20T06:42:25.852663Z"}
{"component":"virt-launcher","kind":"","level":"info","msg":"Domain unpaused.","name":"rke2-2-pool1-926e1b37-fx2bl","namespace":"default","pos":"manager.go:901","timestamp":
"2023-10-20T06:42:25.868221Z","uid":"feeaed37-53a7-4c61-bedd-d56b2f57e61e"}
{"component":"virt-launcher","kind":"","level":"info","msg":"Synced vmi","name":"rke2-2-pool1-926e1b37-fx2bl","namespace":"default","pos":"server.go:190","timestamp":"2023-1
0-20T06:42:25.870412Z","uid":"feeaed37-53a7-4c61-bedd-d56b2f57e61e"}
{"component":"virt-launcher","level":"info","msg":"Domain name event: default_rke2-2-pool1-926e1b37-fx2bl","pos":"client.go:413","timestamp":"2023-10-20T06:42:25.929426Z"}
{"component":"virt-launcher","level":"info","msg":"DomainLifecycle event 4 with reason 0 received","pos":"client.go:435","timestamp":"2023-10-20T06:42:25.939416Z"}
{"component":"virt-launcher","level":"info","msg":"DomainLifecycle event 3 with reason 2 received","pos":"client.go:435","timestamp":"2023-10-20T06:42:25.947074Z"}
{"component":"virt-launcher","level":"info","msg":"kubevirt domain status: Paused(3):IOError(5)","pos":"client.go:288","timestamp":"2023-10-20T06:42:25.947162Z"}
{"component":"virt-launcher","level":"info","msg":"Libvirt event channel is full, dropping event.","pos":"client.go:443","timestamp":"2023-10-20T06:42:25.947174Z"}
{"component":"virt-launcher","kind":"","level":"info","msg":"Domain unpaused.","name":"rke2-2-pool1-926e1b37-fx2bl","namespace":"default","pos":"manager.go:901","timestamp":
"2023-10-20T06:42:25.954738Z","uid":"feeaed37-53a7-4c61-bedd-d56b2f57e61e"}
```

If we run `kg vm rke2-2-pool1-926e1b37-fx2bl -o jsonpath='{.status.printableStatus}' -w`, it changes between Paused and Running.

```
Paused
Running
Running
Paused
Paused
```

The bug is similar to [VM status flipping between Paused and Running](https://bugzilla.redhat.com/show_bug.cgi?id=2029391).

**Comment by FrankYang0529** [comment]
Previous support bundle is based on harvester v1.2.0. This one is based on v1.2.1-rc1. The behavior is same as previous one.

[supportbundle_e6f9e2ea-cdcc-4703-8aff-4157f3646dba_2023-10-24T06-02-42Z.zip](https://github.com/harvester/harvester/files/13111297/supportbundle_e6f9e2ea-cdcc-4703-8aff-4157f3646dba_2023-10-24T06-02-42Z.zip)

**Comment by ejweber** [comment]
In the recreate I analyzed for https://github.com/longhorn/longhorn/issues/6961, the high level issue was that Longhorn intentionally crashed the volume underneath the VM. The VM experienced an I/O error when it tried to write to the crashed volume. Not sure if the exact lead up to the crash is always the same (need to check all the support bundles for that). Longhorn automatically reattaches the volume (which is why it looks good), but it is not usable without restarting the workload.

We should definitely work towards a fix for this on the Longhorn side, though we may need the virtualization layer to change its behavior a bit as well.

I think it may still be a good idea to implement https://github.com/harvester/harvester/pull/4645. Longhorn automatically restarts workload pods when their corresponding volumes have crashed (if the setting is enabled) to avoid similar situations where a workload is stuck indefinitely. The logic is essentially, since there's no other way to recover, we may as well preemptively take the recovery action. (I know there's a PR comment about being too silent when doing this, which I think is also valid.)

**Comment by w13915984028** [comment]
@bk201 @FrankYang0529 

Given the decision https://github.com/harvester/harvester/pull/4645#issuecomment-1777299500, would suggest we add this issue and workaround into Harvester doc (https://docs.harvesterhci.io/v1.3/troubleshooting/index with a new chapter like `Virtual Machine`) as well. thanks.

**Comment by w13915984028** [comment]
@FrankYang0529 @ejweber

Happend to filter this issue https://github.com/harvester/harvester/issues/1905 in context of https://github.com/harvester/harvester/issues/4659.

In #1905 , the VM re-balancing looked caused by others, but the failure

 > but the other 3 existing vms are now blinking (paused/running)

 seemed to be a bit like here #4633.

**Comment by FrankYang0529** [comment]
In this support bundle, there are 2 running engine and VMI can be back without restart.

[supportbundle_e15b5165-f184-45fc-a9ea-02224b33c0de_2023-10-27T02-59-47Z.zip](https://github.com/harvester/harvester/files/13184622/supportbundle_e15b5165-f184-45fc-a9ea-02224b33c0de_2023-10-27T02-59-47Z.zip)

**Comment by FrankYang0529** [comment]
> Given the decision [#4645 (comment)](https://github.com/harvester/harvester/pull/4645#issuecomment-1777299500), would suggest we add this issue and workaround into Harvester doc (https://docs.harvesterhci.io/v1.3/troubleshooting/index with a new chapter like `Virtual Machine`) as well. thanks.

Sounds good. I will add it.

**Comment by FrankYang0529** [comment]
Wait for LH v1.4.5 and bump it in Harv v1.2.2.

**Comment by ejweber** [comment]
IMO Longhorn v1.4.5 will not be enough.

Longhorn v1.4.5 will mitigate against the I/O errors in this issue by fixing https://github.com/longhorn/longhorn/issues/6961. However, there is an underlying issue in Kubernetes that will keep Longhorn volumes attached to two nodes indefinitely after unplanned reboots. In most cases https://github.com/longhorn/longhorn/issues/6961 will keep the Longhorn engine alive on the new node the VM has migrated to, but we will observe a number of additional issues until the volume is detached from the old node.

This behavior will be entirely fixed upstream by https://github.com/kubernetes/kubernetes/pull/122211 (tracked in https://github.com/longhorn/longhorn/issues/6919), but we will need a new Kubernetes version to make use of it.

The workaround for https://github.com/longhorn/longhorn/issues/6919 is to restart the kubelet after the node has come back online and is logging the following:

```
E1019 01:11:18.316662     967 nestedpendingoperations.go:348] Operation for "{volumeName:kubernetes.io/csi/driver.longhorn.io^pvc-37461bdd-7133-4111-8fb7-c4ddf8dacca5 podName:19e41dfe-8cee-40c2-a39b-37c68b01c9a7 nodeName:}" failed. No retries permitted until 2023-10-19 01:13:20.316609446 +0000 UTC m=+1883.799156551 (durationBeforeRetry 2m2s). Error: UnmapVolume.UnmapBlockVolume failed for volume "pvc-37461bdd-7133-4111-8fb7-c4ddf8dacca5" (UniqueName: "kubernetes.io/csi/driver.longhorn.io^pvc-37461bdd-7133-4111-8fb7-c4ddf8dacca5") pod "19e41dfe-8cee-40c2-a39b-37c68b01c9a7" (UID: "19e41dfe-8cee-40c2-a39b-37c68b01c9a7") : blkUtil.DetachFileDevice failed. globalUnmapPath:, podUID: 19e41dfe-8cee-40c2-a39b-37c68b01c9a7, bindMount: true: failed to unmap device from map path. mapPath is empty
```

This was not possible before without causing further I/O errors, but https://github.com/longhorn/longhorn/issues/7302 (which will also be fixed in v1.4.5) allows for it.

In summary, Harvester needs at least Longhorn v1.4.5, v1.5.4, or v1.6.0 AND one of the following:

- A Kubernetes version with https://github.com/kubernetes/kubernetes/pull/122211, or
- A mechanism by which the Kubelet is restarted when it starts logging the above.

cc @phanle1010

**Comment by w13915984028** [comment]
@FrankYang0529 
Could we double check, with bumped LH, k8s version (>=v1.2.6 release) in Harvester v1.3.0, this issue has been fixed?

LH fix: from v1.4.x
Fix engine migration crash (backport #2275) https://github.com/longhorn/longhorn-manager/pull/2287

K8s fix from v1.26
Automated cherry pick of #122211: Fix device uncertain errors on reboot https://github.com/kubernetes/kubernetes/pull/122398


cc @bk201 @markhillgit : should we move this issue to v1.3.0 to validate, and add backport v1.2.2 ?

BTW: the fix of this issue, is a very important enhancement of https://github.com/harvester/harvester/issues/3261

**Comment by markhillgit** [comment]
Please move to ready for testing if appropriate, I've optimistically moved this into 1.3.0 with 1.2.2 backport per the comments above

**Comment by FrankYang0529** [comment]
Currently, we are still waiting for RKE2 release. Both v1.26.13 and v1.27.10 are pre-released.

* https://github.com/rancher/rke2/releases/tag/v1.26.13-rc2%2Brke2r1
* https://github.com/rancher/rke2/releases/tag/v1.27.10-rc2%2Brke2r1

**Comment by w13915984028** [comment]
RKE2 v1.27.10+rke2r1 has been released
https://github.com/rancher/rke2/releases/tag/v1.27.10%2Brke2r1

It has been merged to Harvester in https://github.com/harvester/harvester-installer/pull/655

**Comment by w13915984028** [comment]
To QA:

When validating this issue, please note those issues are caused by same root cause though they have slightly different phenomenon.
#4633 
#4033
#3539
#4789

**Comment by albinsun** [comment]
Provide test result on `v1.3.0-rc3`, close as fixed.


## Environment
* Harvester
  * Version: `v1.3.0-rc3`
  * Profile: QEMU/KVM, 3 nodes (8C/16G/500G)
  * ui-source: `Auto`


## Steps 
1. Create a VM with LVM
   * <details>
     <summary> Mount LVM and generate some test data :heavy_check_mark: 
     </summary>

     ![image](https://github.com/harvester/harvester/assets/2773781/32e6eede-963a-4b9f-b86a-4715ed440f93)
     </details>

   
1. Restart the node VM on (1st try)
   * <details>
     <summary> Freeze kubelet and node is <code>Unavailable</code>, then restart node :heavy_check_mark: 
     </summary>
     
     ![image](https://github.com/harvester/harvester/assets/2773781/ecc73b67-7664-4e43-96db-a02b036a3bb2)
     </details>
   * <details>
     <summary> VM becomes <code>Not Ready</code> :heavy_check_mark: 
     </summary>
     
     ![image](https://github.com/harvester/harvester/assets/2773781/d42fa0d3-7711-45f7-bb53-11123db50dcb)
     </details>
   * <details>
     <summary> Node is back and VM can be <code>Running</code> again after a while :heavy_check_mark: 
     </summary>
     
     ![image](https://github.com/harvester/harvester/assets/2773781/431e79b4-f02b-40ac-bce3-27e62c20de6d)
     ![image](https://github.com/harvester/harvester/assets/2773781/6ba84a68-c28e-4311-8312-bcaeac45d835)
     </details>


1. Restart the node VM on (2nd try)
   * <details>
     <summary> Freeze kubelet and node is <code>Unavailable</code>, then restart node :heavy_check_mark: 
     </summary>
     
     ![image](https://github.com/harvester/harvester/assets/2773781/d4d3286c-c691-4034-ace4-3d63bb3f16cc)
     </details>
   * <details>
     <summary> VM becomes <code>Not Ready</code> :heavy_check_mark: 
     </summary>
     
     ![image](https://github.com/harvester/harvester/assets/2773781/a466abf3-7e76-47d5-a603-96a5a967ecf5)
     </details>
   * <details>
     <summary> Node is back and VM can be <code>Running</code> again after a while :heavy_check_mark: 
     </summary>
     
     ![image](https://github.com/harvester/harvester/assets/2773781/b02cbeb1-81d4-4cf8-867b-e4ecffe9837f)

     ![image](https://github.com/harvester/harvester/assets/2773781/f4255c32-e00e-4fb5-a94c-0ed336f47f87)
     </details>


1. Restart the node VM on (3rd try)
   * <details>
     <summary> Freeze kubelet and node is <code>Unavailable</code>, then restart node :heavy_check_mark: 
     </summary>
     
     ![image](https://github.com/harvester/harvester/assets/2773781/65cd7ff4-e9f0-4280-8a2c-d7747c36c00b)
     </details>
   * <details>
     <summary> VM becomes <code>Not Ready</code> :heavy_check_mark: 
     </summary>
     
     ![image](https://github.com/harvester/harvester/assets/2773781/769a88fd-e75b-4d9d-8e2d-f61410ae5a15)
     </details>
   * <details>
     <summary> Node is back and VM can be <code>Running</code> again after a while :heavy_check_mark: 
     </summary>
     
     ![image](https://github.com/harvester/harvester/assets/2773781/3836e89a-27e8-416d-b51d-b18844a424b0)
     ![image](https://github.com/harvester/harvester/assets/2773781/d0cdb634-54d9-4963-8f04-e1188c34a3b7)
     </details>


## Note
1. Can be reproduce on `v1.2.1`
   <details>

   ![image](https://github.com/harvester/harvester/assets/2773781/7eb31eff-1b07-4f3b-b1e2-6f826599471d)
   Host is back, but VM still NotReady
   ![image](https://github.com/harvester/harvester/assets/2773781/4cd9dd72-1451-469f-8e32-5e7b4f43c99e)
   ![image](https://github.com/harvester/harvester/assets/2773781/704834cd-149c-45cb-a90e-49979ce09e5d)
   </details>
